File: src/autogenerated/java/com/aznos/data/DimensionTypes.java
package com.aznos.data;

import java.lang.String;
import org.jetbrains.annotations.NotNull;

public enum DimensionTypes {
  THE_NETHER(0.1,false,8.0,"minecraft:the_nether",true,false,false,256,"#minecraft:infiniburn_nether",128,0,false,true,true,true),

  OVERWORLD(0.0,true,1.0,"minecraft:overworld",false,true,true,384,"#minecraft:infiniburn_overworld",384,-64,true,false,false,false),

  THE_END(0.0,false,1.0,"minecraft:the_end",false,true,false,256,"#minecraft:infiniburn_end",256,0,false,false,false,false);

  public final float ambientLight;

  public final boolean bedWorks;

  public final double coordinateScale;

  public final String effects;

  public final boolean hasCeiling;

  public final boolean hasRaids;

  public final boolean hasSkyLight;

  public final int height;

  public final String infiniburn;

  public final int logicalHeight;

  public final int minY;

  public final boolean natural;

  public final boolean piglinSafe;

  public final boolean respawnAnchorWorks;

  public final boolean ultrawarm;

  DimensionTypes(@NotNull double ambientLight, @NotNull boolean bedWorks,
      @NotNull double coordinateScale, @NotNull String effects, @NotNull boolean hasCeiling,
      @NotNull boolean hasRaids, @NotNull boolean hasSkyLight, @NotNull int height,
      @NotNull String infiniburn, @NotNull int logicalHeight, @NotNull int minY,
      @NotNull boolean natural, @NotNull boolean piglinSafe, @NotNull boolean respawnAnchorWorks,
      @NotNull boolean ultrawarm) {
    this.ambientLight = (float) ambientLight;
    this.bedWorks = bedWorks;
    this.coordinateScale = coordinateScale;
    this.effects = effects;
    this.hasCeiling = hasCeiling;
    this.hasRaids = hasRaids;
    this.hasSkyLight = hasSkyLight;
    this.height = height;
    this.infiniburn = infiniburn;
    this.logicalHeight = logicalHeight;
    this.minY = minY;
    this.natural = natural;
    this.piglinSafe = piglinSafe;
    this.respawnAnchorWorks = respawnAnchorWorks;
    this.ultrawarm = ultrawarm;
  }
}

File: src/main/kotlin/registry/DimensionTypes.kt
package com.aznos.registry

import com.aznos.datatypes.putBoolean
import dev.dewy.nbt.tags.collection.CompoundTag

class DimensionTypes : Registry<DimensionType>("minecraft:dimension_type") {

    val overworld = register("minecraft:overworld", DimensionType(
        true,
        true,
        -64,
        384,
        "minecraft:overworld",
        0f,
        false
    ))

    val nether = register("minecraft:the_nether", DimensionType(
        false,
        false,
        0,
        256,
        "minecraft:the_nether",
        0.1f,
        true
    ))

    val end = register("minecraft:the_end", DimensionType(
        false,
        false,
        0,
        256,
        "minecraft:the_end",
        0f,
        false
    )
    )

    override fun asCompound(value: DimensionType): CompoundTag {
        val result = CompoundTag()

        result.putBoolean("has_skylight", value.hasSkylight)
        result.putBoolean("natural", value.natural)
        result.putInt("min_y", value.minY)
        result.putInt("height", value.height)
        result.putString("effects", value.effects)
        result.putFloat("ambient_light", value.ambientLight)
        result.putBoolean("piglin_safe", value.piglinSafe)

        // These have no effect on the client so it doesn't matter what it is
        result.putBoolean("has_ceiling", false)
        result.putBoolean("ultrawarm", false)
        result.putDouble("coordinate_scale", 1.0)
        result.putBoolean("bed_works", false)
        result.putBoolean("respawn_anchor_works", false)
        result.putInt("logical_height", value.height - 1)
        result.putString("infiniburn", "#minecraft:overworld")
        result.putBoolean("has_raids", false)
        result.putInt("monster_spawn_light_level", 0)
        result.putInt("monster_spawn_block_light_limit", 0)

        return result
    }

}

data class DimensionType( // Left out the attributes that has no effect on the client
    val hasSkylight: Boolean,
    val natural: Boolean,
    val minY: Int,
    val height: Int,
    val effects: String,
    val ambientLight: Float,
    val piglinSafe: Boolean
)